packages/terminal/src/browser/base/terminal-client.ts

```diff
@@ -0,0 +1,98 @@
+/********************************************************************************
+ * Copyright (C) 2019 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { IShellTerminalServerOptions } from '../../common/shell-terminal-protocol';
+import { Disposable } from '@theia/core';
+import { TerminalWidget, TerminalSize } from './terminal-widget';
+
+export type TerminalClientFactory = (options: TerminalClientOptions, terminalWidget: TerminalWidget) => TerminalClient;
+export const TerminalClientFactory = Symbol('TerminalClientFactory');
+
+/**
+ * TerminalClient contains connection logic between terminal server side and terminal widget.
+ * TerminalClient is responsible:
+ * - create new terminal process;
+ * - attach to the already running process on the backend(in case if backend support it);
+ * - create connection with backend process;
+ * - send output to the terminal widget from backend;
+ * - send user input from terminal widget to the backend.
+ * - kill backend process(in case if backend support it);
+ * - resize backend output frame(in case if backend support it);
+ */
+export const TerminalClient = Symbol('TerminalClient');
+export interface TerminalClient extends Disposable {
+
+    /**
+     * Terminal client options to setup backend terminal process
+     * and control terminal client configuration.
+     */
+    readonly options: TerminalClientOptions;
+
+    /**
+     * Terminal widget which is under TerminalClient control.
+     */
+    readonly widget: TerminalWidget;
```

Comment:
that's very weird that there is a cycle dependency between `TerminalClient` and `TerminalWidget`